### PR TITLE
Make retrying transport and http errors configurable 

### PR DESCRIPTION
**Pull request by DennisDenuto**
Related to https://github.com/google/go-containerregistry/issues/1114

Adds the following 4 options:
- WithRetryHTTPBackoff
- WithRetryHTTPPredicate
- WithRetryTransportBackoff
- WithRetryTransportPredicate

Went with a finer level API around configuring these (in my head) 2 groups of retry logic (low-level transport errors and higher level HTTP errors).

I was also aiming to keep the 'default' retry logic values the same (if these options are not provided).

### CODE REVIEW

**Comment by jonjohnsonjr** [comment]
So I don't love how many options this adds, but I get the need for this to be more configurable.

For the transport pieces, I think I'd like to revisit this: https://github.com/google/go-containerregistry/issues/740

Ideally, you could supply your own transport via `remote.WithTransport` and use `transport.NewRetry` to configure it. The only issue is around double-wrapping because of how we wrap transports by default. One option here is to just not wrap the supplied transport if you pass in a custom transport, but then you'd lose out on some "niceness". If we exposed [this stuff](https://github.com/google/go-containerregistry/blob/8388fde8bcadbc8ddb820b5ac0046e10e6c572e7/pkg/v1/remote/options.go#L81-L94) in a nice way, you could claw that back by opting into it, e.g.:

```go
// Wraps with retries and useragent and debug logging
remote.Image(..., remote.WithTransport(remote.Transport{t}))

// No wrapping of t
remote.Image(..., remote.WithTransport(t))

// Configurable backoff stuff.
remote.Image(..., remote.WithTransport(transport.NewRetry(t, ...)))
```

My only issue with that approach is that it's probably a breaking change :/

Another approach might be exposing more on the errors we return from `transport`, e.g. the request: https://github.com/google/go-containerregistry/blob/8388fde8bcadbc8ddb820b5ac0046e10e6c572e7/pkg/v1/remote/transport/error.go#L52

That would allow you to write a predicate that checked the request method, URL, and status code yourself to determine if you should retry it.

**Comment by DennisDenuto** [comment]
@jonjohnsonjr 

Yeah I think iterating over the API change to make it as minimal as needed makes lots of sense. Both from maintainability and from a usability perspective! With the changes above sounds like a trade-off between a simpler API vs breaking changes to existing APIs.

I played around with the idea of using `WithTransport` as a way to allow a consumer to configure retrying.

> Another approach might be exposing more on the errors we return from transport

I did this by creating a new `RetryError` that wraps the transport error (if it exists), and exposes the status code and request.
I also made it implement the `Temporary` function to make it a compatible with the ['default' transport predicate](https://github.com/google/go-containerregistry/blob/main/pkg/v1/remote/transport/retry.go#L67)

I pushed the changes. I'd love some early feedback to make sure i'm properly understanding your idea to begin with!

This PR comes with breaking changes:
- Using `WithTransport(t)` with a non `remote.Transport` no longer wraps retry, logging etc...
- Retrying with the transport.Retry will return a `RetryError` (even if the `RoundTripper` did not return an error). It always returns a `RetryError` since it contains http data (status code) to allow specifying a `Predicate` to determine whether to retry or not.

Note:

Some tests are failing due to these breaking changes. But I wanted to get feedback before proceeding down this path.

**Comment by jonjohnsonjr** [comment]
I don't think we need `RetryError` if we just expose the existing request field on the `transport.Error` struct. Callers can check if an error is `transport.Error` and use that information. If it's some other kind of error, they can just do that in an "else" rather than using `Inner`.

> Using WithTransport(t) with a non remote.Transport no longer wraps retry, logging etc...

This feels kinda backwards. I think if someone gives us a `transport.Transport`, we should avoid wrapping it at all. We could use that as a signal that the caller knows what they're doing and we should get out of their way.

It would still make sense to me to expose Backoff/Predicate options in `remote` for higher level things that cannot be retried at a request level, e.g. layer uploads. I'm not sure how I would want to structure those hooks, but there are a handful of places that might make sense to inject some kind of retry check.

I'd expect the default predicate to be somewhat conservative, but having the option would let callers do some pretty flexible things like only retry for certain methods or for certain paths.

**Comment by DennisDenuto** [comment]
> if we just expose the existing request field on the transport.Error

hmmm. I don't see how `transport.Error` ever gets returned by a RoundTripper. I *can* however see how it is returned at the http req/resp layer via `CheckError(...)`  

or put another way...

I don't see how a consumer providing their own Retry Transport would ever get an error of type `transport.Error`

Although I see how exposing the `Request` on the `transport.Error` can be useful in conjunction with exposing Backoff/Predicate options in remote for higher level http retries.

> I think if someone gives us a transport.Transport, we should avoid wrapping it at all.

hmm ok, I mis-understood your code snippet above then

```
// Wraps with retries and useragent and debug logging
remote.Image(..., remote.WithTransport(remote.Transport{t}))
```

but thinking about it more, it does make sense that it should *not* wrap, and i think will result in this not being a breaking change. i'll def update the PR with this.

 > It would still make sense to me to expose Backoff/Predicate options in `remote`

I think this helps clarify things a lot for me.

I was trying to have `transport.Retry` provide a way to retry both low level transport errors and http level errors. I updated the PR to put back Backoff/Predicate options in remote :-)

**Comment by jonjohnsonjr** [comment]
> hmm ok, I mis-understood your code snippet above then

Yeah I _didn't_ love my code snippet, but what you've got here is actually a better solution, I think.

> I think this helps clarify things a lot for me.
> 
> I was trying to have transport.Retry provide a way to retry both low level transport errors and http level errors. I updated the PR to put back Backoff/Predicate options in remote :-)

Ah yeah, sorry this wasn't more clear. There were two categories of retry stuff going on, but we kind of lumped them together. Current PR is pretty close to optimal, I think.

File: pkg/v1/remote/transport/transport.go [hunk 1]
```diff
@@ -101,3 +101,14 @@ func NewWithContext(ctx context.Context, reg name.Registry, auth authn.Authentic
 		return nil, fmt.Errorf("unrecognized challenge: %s", pr.challenge)
 	}
 }
+
+// Transport results in *not* wrapping supplied transport with additional logic such as retries, useragent and debug logging
+// Consumers are opt-ing into providing their own transport without any additional wrapping.
+type Transport struct {
```
Comment (jonjohnsonjr) [inline pkg/v1/remote/transport/transport.go:107]: I think all that's left is to make the `transport` package actually return these. 

I don't love the `transport.Transport` name. Maybe `transport.Wrapper` or something?

Another issue is that it's kind of a pain to actually construct a transport outside of the remote package (you need to know where layers came from, where you're pushing, etc). We could make that a lot more ergonomic by introducing a new transport implementation that defers those decisions until they're needed, like containerd (https://github.com/google/go-containerregistry/issues/666#issuecomment-580034491).

File: pkg/v1/remote/options.go [hunk 2]
```diff
@@ -78,19 +107,21 @@ func makeOptions(target authn.Resource, opts ...Option) (*options, error) {
 		o.auth = auth
 	}
 
-	// Wrap the transport in something that logs requests and responses.
-	// It's expensive to generate the dumps, so skip it if we're writing
-	// to nothing.
-	if logs.Enabled(logs.Debug) {
-		o.transport = transport.NewLogger(o.transport)
-	}
+	if _, ok := o.transport.(*transport.Transport); !ok {
```
Comment (jonjohnsonjr) [inline pkg/v1/remote/options.go:110]: Add a comment here and in `google` describing _why_ we're doing things this way.

Also for the `WithTransport` option, we'll want to describe this workaround.

Comment (DennisDenuto) [reply pkg/v1/remote/options.go]: done!

File: pkg/v1/remote/transport/transport.go (hunk 1 above)
Comment (DennisDenuto) [reply pkg/v1/remote/transport/transport.go]: Yep! changed it to `transport.Wrapper`

>  make the transport package actually return these.

I wasn't 100% sure what you meant here. I took this to mean `NewWithContext` should return `transport.Wrapper`. I pushed the change.

> Another issue is that it's kind of a pain to actually construct a transport outside of the remote package (you need to know where layers came from, where you're pushing, etc). We could make that a lot more ergonomic by introducing a new transport implementation that defers those decisions until they're needed, like containerd (#666 (comment)).

hmm ok, is it possible to split introducing a new deferring transport implementation into a separate PR? Or do you see this as a blocker to this PR?

Comment (jonjohnsonjr) [reply pkg/v1/remote/transport/transport.go]: > I wasn't 100% sure what you meant here. I took this to mean NewWithContext should return transport.Wrapper. I pushed the change.

Yep, that's perfect.

> hmm ok, is it possible to split introducing a new deferring transport implementation into a separate PR? Or do you see this as a blocker to this PR?

Not a blocker, just thinking out loud.

File: pkg/v1/remote/options.go [hunk 3]
```diff
@@ -78,26 +107,32 @@ func makeOptions(target authn.Resource, opts ...Option) (*options, error) {
...
 
-	// Wrap the transport in something that logs requests and responses.
-	// It's expensive to generate the dumps, so skip it if we're writing
-	// to nothing.
-	if logs.Enabled(logs.Debug) {
-		o.transport = transport.NewLogger(o.transport)
-	}
+	// transport.Wrapper is a signal that consumers are opt-ing into providing their own transport without any additional wrapping.
+	// This is to allow consumers full control over the transports logic, such as providing retry logic.
+	if _, ok := o.transport.(*transport.Wrapper); !ok {
```
Comment (jonjohnsonjr) [inline pkg/v1/remote/options.go:112]: This is going to be tricky to maintain -- can you add a test that _doesn't_ get wrapped so we know this works?

Comment (DennisDenuto) [reply pkg/v1/remote/options.go:112]: Finding a good place to test this wasn't a clear call.

Ended up writing a couple tests in the `MultiWrite` test:
-  Asserts that if a `transport.Wrapper` is used, then retrying should *not* occur. 
-  Asserts that if a `transport.Wrapper` is used, then User-Agent should not be wrapped..

**Comment by DennisDenuto** [comment]
@jonjohnsonjr Just checking in to see if theres anything else needed to have this merged in?

**Comment by jonjohnsonjr** [comment]
Sorry I missed your follow-up commits!

One final thing I forgot to mention. If we're dealing with a `transport.Wrapper`, we probably also want to skip all of this:
https://github.com/google/go-containerregistry/blob/8388fde8bcadbc8ddb820b5ac0046e10e6c572e7/pkg/v1/remote/transport/transport.go#L53-L68

We can do a little introspection to determine if the the given transport can be reused by looking at `reg` to see if it's the same as what `t` already pinged. Then we can skip some pinging. For the bearer case, we probably want to merge the existing scopes with the given `scopes` like with: https://github.com/google/go-containerregistry/blob/8388fde8bcadbc8ddb820b5ac0046e10e6c572e7/pkg/v1/remote/transport/bearer.go#L99-L105

**Comment by DennisDenuto** [comment]
Thanks I appreciate the pointers @jonjohnsonjr 

> We can do a little introspection to determine if the the given transport can be reused by looking at reg to see if it's the same as what t already pinged

I did this by saving state ('pingedRegistries') in `transport.Wrapper`

> For the bearer case, we probably want to merge the existing scopes with the given scopes like with:

I figured that the scopes should be tied to the registry that the transport pinged against. Hence scopes being a map on the wrapped transport.

**Comment by jonjohnsonjr** [comment]
Okay so this might make you hate me, but I'm not thrilled with how this turned out after my suggestion (I know... I'm sorry).

Do you mind backing this out to the previous revision? We can just submit because it's useful on its own, then I can revisit the ping optimizations when I have more time to look at it.

Again, sorry, I appreciate your patience.

**Comment by DennisDenuto** [comment]
@jonjohnsonjr  lol no worries. i've reverted that commit

fwiw I have enjoyed working on this PR and we get a ton of value from using this library. so more than happy to be patient and get things done right! 😄

File: pkg/v1/remote/options.go [hunk 4]
```diff
@@ -78,26 +107,32 @@ func makeOptions(target authn.Resource, opts ...Option) (*options, error) {
...
+		// It's expensive to generate the dumps, so skip it if we're writing
+		// to nothing.
+		if logs.Enabled(logs.Debug) {
+			o.transport = transport.NewLogger(o.transport)
+		}
 
-	// Wrap the transport in something that can retry network flakes.
-	o.transport = transport.NewRetry(o.transport)
+		// Wrap the transport in something that can retry network flakes.
+		o.transport = transport.NewRetry(o.transport, transport.WithRetryBackoff(o.retryBackoff), transport.WithRetryPredicate(o.retryPredicate))
```
Comment (jonjohnsonjr) [inline pkg/v1/remote/options.go:121]: I'm not sure if we want to plumb this down into `transport.NewRetry` -- but maybe you can convince me.

The defaults for that transports are geared towards smaller blips (maybe this makes sense, maybe it doesn't), and I'd expect the `remote` options are more geared towards higher level retries (i.e. at the HTTP level).

File: pkg/v1/remote/transport/transport.go [hunk 5]
```diff
@@ -96,8 +96,19 @@ func NewWithContext(ctx context.Context, reg name.Registry, auth authn.Authentic
...
+		return &Wrapper{bt}, nil
 	default:
 		return nil, fmt.Errorf("unrecognized challenge: %s", pr.challenge)
 	}
 }
+
+// Wrapper results in *not* wrapping supplied transport with additional logic such as retries, useragent and debug logging
+// Consumers are opt-ing into providing their own transport without any additional wrapping.
+type Wrapper struct {
+	Inner http.RoundTripper
```
Comment (jonjohnsonjr) [inline pkg/v1/remote/transport/transport.go:108]: Let's not expose `Inner` quite yet? I think we only want you to be able to produce a `Wrapper` by going through `transport.New` for now. We can relax that later if it makes sense.

Comment (DennisDenuto) [reply pkg/v1/remote/transport/transport.go]: ok done!

**Review by jonjohnsonjr — APPROVED** [review]
Thanks for sticking with this. This looks good to me -- let's merge and let it bake for a bit while I try to figure out how we can get rid of extra pings :)

**Pull request merged**